firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,301 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+    use InnerNode::*;
+
+    let Some((visited_parent, visited_pos)) = visited_parents.pop() else {
+        return Ok(None);
+    };
+
+    let mut node = ParentNode::Visited(visited_parent);
+    let mut pos = visited_pos;
+    let mut first_loop = true;
+
+    loop {
+        match node.inner() {
+            New(NodeType::Leaf(leaf)) => {
+                let value = leaf.data.to_vec();
+                return Ok(Some((node.into_node(), value)));
+            }
+
+            Visited(NodeType::Leaf(_)) | Visited(NodeType::Extension(_)) => {
+                let Some((next_parent, next_pos)) = visited_parents.pop() else {
+                    return Ok(None);
                 };
 
-                (found_key, value)
+                node = ParentNode::Visited(next_parent);
+                pos = next_pos;
             }
-            _ => unreachable!(),
-        };
 
-        Poll::Ready(Some(Ok(return_value)))
+            New(NodeType::Extension(extension)) => {
+                let child = merkle.get_node(extension.chd())?;
+
+                pos = 0;
+                visited_parents.push((node.into_node(), pos));
+
+                node = ParentNode::New(child);
+            }
+
+            Visited(NodeType::Branch(branch)) => {
+                let compare_op = if first_loop {
+                    <u8 as PartialOrd>::ge
+                } else {
+                    <u8 as PartialOrd>::gt
+                };
+
+                let children = get_children_iter(branch)
+                    .filter(move |(_, child_pos)| compare_op(child_pos, &pos));
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+
+            New(NodeType::Branch(branch)) => {
+                if let Some(value) = branch.value.as_ref() {
+                    let value = value.to_vec();
+                    return Ok(Some((node.into_node(), value)));
+                }
+
+                let children = get_children_iter(branch);
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+        }
+
+        first_loop = false;
     }
 }
 
-/// Compute a key from a set of parents
-fn key_from_parents(parents: &[(ObjRef, u8)]) -> Vec<u8> {
-    parents[1..]
-        .chunks_exact(2)
-        .map(|parents| (parents[0].1 << 4) + parents[1].1)
-        .collect::<Vec<u8>>()
+fn get_children_iter(branch: &BranchNode) -> impl Iterator<Item = (DiskAddress, u8)> {
+    branch
+        .children
+        .into_iter()
+        .enumerate()
+        .filter_map(|(pos, child_addr)| child_addr.map(|child_addr| (child_addr, pos as u8)))
+}
+
+#[must_use]
```

Comment:
A1: yes
A2: I can change it to a `bool` if you prefer, it definitely does not need to be generic
A3: I put zero thought into its location. Happy to move it.

To be honest, I'm not a huge fan of this code, but it works, and it prevents someone from ignoring the result and forgetting to early-return. I have some ideas on how I could improve the ergonomics here, but great is the enemy of good enough, so I've been holding off.

Reply:
Bool would be better. This could use some doc comments as well, particularly around why you're doing this and why it's important in this particular case. Rust doesn't have a good way to force a must use of the inner part of the result.

Reply:
I moved both `Either` and `MustUse` to a sub-mod within the same file called `helper_types`. I left `MustUse` as generic because the `#[must_use]` attribute itself is generic. When I read
```rust
#[must_use]
struct MustUse(bool) 
```
I ask myself what the purpose of the `bool` is. Whereas with the generic type, I know that I have some wrapper that I must do *something* with. It's also obvious that it only serves *one* purpose since `T` could be anything including `()`.